src/systems/python_system_loader/PythonSystemLoader.cc

```diff
@@ -0,0 +1,198 @@
+    }
+  }
+  catch (const std::runtime_error &_err)
+  {
+    gzerr << _err.what() << std::endl;
+    return;
+  }
+
+  try
+  {
+    this->pythonModule = py::module_::import(moduleName.c_str());
+  }
+  catch (const pybind11::error_already_set &_err)
+  {
+    gzerr << "Error while loading module " << moduleName << "\n"
+          << _err.what() << "\n";
+
+    gzerr << "Searched in:\n";
+    for (const auto &path : sysModule.attr("path"))
+    {
+      gzerr << "  \"" << path << "\"\n";
+    }
+    return;
+  }
+
+  try
+  {
+    py::object getSystem = this->pythonModule.attr("get_system");
+    if (!getSystem)
+    {
+      gzerr << "Could not call 'get_system' on the module" << moduleName << "\n";
+      return;
+    }
+    this->pythonSystem = getSystem();
+  }
+  catch (const pybind11::error_already_set &_err)
+  {
+    gzerr << _err.what() << std::endl;
+    return;
+  }
+  if (py::hasattr(this->pythonSystem, "configure"))
+  {
+    try
+    {
+      // _ecm and _eventMgr are passed in as pointers so pybind11 would use the
+      // write `return_value_policy`. Otherwise, it would use
+      // `return_value_policy = copy`.
+      this->pythonSystem.attr("configure")(_entity, _sdf, &_ecm, &_eventMgr);
+    }
+    catch (const pybind11::error_already_set &_err)
+    {
+      gzerr << _err.what() << std::endl;
+      return;
+    }
+  }
+  // TODO(azeey) Add support for ConfigureParameters
+  if (py::hasattr(this->pythonSystem, "pre_update"))
+  {
+    this->preUpdateMethod = this->pythonSystem.attr("pre_update");
+  }
+  if (py::hasattr(this->pythonSystem, "update"))
+  {
+    this->updateMethod = this->pythonSystem.attr("update");
+  }
+  if (py::hasattr(this->pythonSystem, "post_update"))
+  {
+    this->postUpdateMethod = this->pythonSystem.attr("post_update");
+  }
+  if (py::hasattr(this->pythonSystem, "reset"))
+  {
+    this->resetMethod = this->pythonSystem.attr("reset");
+  }
+  this->validConfig = true;
+}
+
+//////////////////////////////////////////////////
+template <typename ...Args>
+void PythonSystemLoader::CallPythonMethod(py::object _method, Args&&... _args)
+{
+  if (!this->validConfig)
+  {
+    return;
+  }
+
+  if (_method)
+  {
+    try
+    {
+      _method(std::forward<Args>(_args)...);
+    }
+    catch (const pybind11::error_already_set &_err)
+    {
+      gzerr << _err.what() << std::endl;
+      return;
+    }
+  }
+}
+
+//////////////////////////////////////////////////
+void PythonSystemLoader::PreUpdate(const UpdateInfo &_info,
+                                   EntityComponentManager &_ecm)
+{
+  CallPythonMethod(this->preUpdateMethod, _info, &_ecm);
+}
+
+//////////////////////////////////////////////////
+void PythonSystemLoader::Update(const UpdateInfo &_info,
+                                EntityComponentManager &_ecm)
+{
+  CallPythonMethod(this->updateMethod, _info, &_ecm);
+}
+
+//////////////////////////////////////////////////
+void PythonSystemLoader::PostUpdate(const UpdateInfo &_info,
```

Comment:
Did we determine what happens when there are multiple python modules?  Is that supported/unsupported?

Postupdate is the most interesting case because of parallel execution

Reply:
So, we're currently creating an instance of ` pybind11::scoped_interpreter` in `ServerPrivate` and it seems to work fine with multiple python modules, since there's only one `Server` instance. However, if we want to have multiple `Server`s in the same process, this will not work. So maybe we need a Singleton for this.

Reply:
I think we can generally assume that multiple Servers per process isn't likely at the moment. We may want to add a note somewhere in the event that someone attempts it, though.

Reply:
Some of the tests were failing because they initialized two `Server` instances. I fixed it by using `pybind11::initialize_interpreter()` instead of `pybind11::scoped_interpreter` so we can skip initialization of the interpreter is already initialized.